File: Medium/515. Find Largest Value in Each Tree Row/515. Find Largest Value in Each Tree Row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        result = []
        queue = deque()
        queue.append(root)

        while queue:
            row_max = float("-inf")  # Initialize with negative infinity
            row_size = len(queue)

            for _ in range(row_size):
                node = queue.popleft()
                row_max = max(row_max, node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(row_max)

        return result

File: Medium/515. Find Largest Value in Each Tree Row/515. Find Largest Value in Each Tree Row.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val)
 *     this.left = (left === undefined ? null : left)
 *     this.right = (right === undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
    if (!root) return [];

    const result = [];
    let queue = [root];

    while (queue.length > 0) {
        let rowMax = Number.MIN_SAFE_INTEGER;
        const rowSize = queue.length;

        for (let i = 0; i < rowSize; i++) {
            const node = queue.shift();
            rowMax = Math.max(rowMax, node.val);
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }

        result.push(rowMax);
    }

    return result;
};
